[custom_op] Make repeated registrations error gracefully

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #101015
* #101190
* #100980
* __->__ #100979

Previously the error message went through torch.library. This PR changes
it so that on each custom_op.impl_* call:
- we store a (function, location) tuple
- if a (function, location) tuple exists already, then we raise an
error.

This logic already existed for the abstract impl (the impl for meta and
fake tensors), so this PR just extends it to the others.

Test Plan:
- new test

zou3519 committed:
[custom_op] Make repeated registrations error gracefully

Previously the error message went through torch.library. This PR changes
it so that on each custom_op.impl_* call:
- we store a (function, location) tuple
- if a (function, location) tuple exists already, then we raise an
error.

This logic already existed for the abstract impl (the impl for meta and
fake tensors), so this PR just extends it to the others.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] Make repeated registrations error gracefully"


Previously the error message went through torch.library. This PR changes
it so that on each custom_op.impl_* call:
- we store a (function, location) tuple
- if a (function, location) tuple exists already, then we raise an
error.

This logic already existed for the abstract impl (the impl for meta and
fake tensors), so this PR just extends it to the others.

Test Plan:
- new test

[ghstack-poisoned]

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -803,6 +803,26 @@ def foo_meta(x, dim):
+            del output_shape[dim]
+            return x.new_empty(output_shape)
+
+        with self.assertRaisesRegex(RuntimeError, "already has a abstract impl"):
```
nit: the error also includes the location of the registration, is it worth including that in the regex? (since grabbing it based on a hardcoded number of stack frames seems prone to drift over time).

zou3519 replied:
Yeah that's a good idea

zou3519 committed:
Update on "[custom_op] Make repeated registrations error gracefully"


Previously the error message went through torch.library. This PR changes
it so that on each custom_op.impl_* call:
- we store a (function, location) tuple
- if a (function, location) tuple exists already, then we raise an
error.

This logic already existed for the abstract impl (the impl for meta and
fake tensors), so this PR just extends it to the others.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] Make repeated registrations error gracefully"


Previously the error message went through torch.library. This PR changes
it so that on each custom_op.impl_* call:
- we store a (function, location) tuple
- if a (function, location) tuple exists already, then we raise an
error.

This logic already existed for the abstract impl (the impl for meta and
fake tensors), so this PR just extends it to the others.

Test Plan:
- new test

[ghstack-poisoned]
